### ISSUE TITLE
Pinning python packages to check against Pylint etc

### ISSUE THREAD

**Issue by leelavg** [kadalu/kadalu#528]
Rationale:
- We are using all python modules inside containers and we have already pinned base container versions
- Whenever github runner updates it'll install all latest available packages as per requirements.txt which fails pylint
- We don't want to stop consuming latest github runner however it's better to pin versions in requirements.txt to almost resemble what's running in container
- We can also pin versions in Dockerfiles as well and raise separate PR for updating dependencies

**Comment by amarts** [comment]
This is Good. Every major release we can look at the latest versions, and have a call on it.

**Comment by slayer321** [comment]
Hey @amarts, I am new to open source and k8.  I will like to contribute to this project  can you guide me what needs to be done.

**Comment by leelavg** [comment]
@slayer321 Welcome,

I'll try to be as concise as possible wrt requirement and a probable solution for a quick head-start and please comment for more info.

Requirement:
1. There are mainly three components being packaged into images with some helper images as well
2. Some of these images may or may not include python packages and for those included python packages we need to refactor them from dockerfile and use a ~requirements.txt
3. For a start we can have one requirements.txt as a base and remaining txt's which adds extra pkgs to this base
4. And before finalizing the pkg versions we need to make sure versions in the image matches what we're using in CI (refer `make pylint`)
5. For now we are sure that pkgs in existing images are working without any issues and you can pull them and query for version from them without breaking any incompatibility

Probable Solution:
1. Images which include pip installed pkgs are as below:
``` sh
$ find . -type f -name *Dockerfile* | grep -Pv 'tests|tie|sample|server|monitoring'
./csi/Dockerfile
./extras/Dockerfile.python-packages
./extras/Dockerfile.builder
./operator/Dockerfile
```
2. You can refer images created from above files at https://hub.docker.com/u/kadalu or if you've an existing k8s cluster you can just install Kadalu and they all will be pulled automatically
3. After you pull them, query the version of pip installed pkg (refer resp. Dockerfile) and take note of it
4. Finally, create a setup.cfg file at the root of the repo with contents similar to below
``` cfg
$ bat setup.cfg --plain
# All packages should be versioned, this is just a sample
[options]
python_requires = >=3.7
install_requires =
  setuptools
  prometheus-client
  jinja2
# .... remaining pkgs in Dockerfile.builder if any

[options.extras_require]
operator = kubernetes==11.0.0 # ... remaining pkgs in operation/Dockerfile if any
csi = googleapis-common-protos # ... remaining pkgs in csi/Dockerfile if any
pypa = pyxattr; grpcio # ... remaining pkgs in extras/Dockerfile.python-packages if any
```
5. Install pip-tools and then just generate dependencies for above using below commands:
``` sh
$ pip-compile -o base-requirements.txt
$ pip-compile --extra=operator -o operator-requirements.txt
$ pip-compile --extra=csi -o csi-requirements.txt
$ pip-compile --extra=pypa -o pypa-requirements.txt
```
6. Now you can edit the respective dockerfiles to install pkgs using above created txt files, refer [this article](https://pythonspeed.com/articles/dockerizing-python-is-hard/)
7. Once all is in place edit below line to include above created txt files and remove existing requirements.txt file
``` sh
$ sed -n 47p .github/workflows/on-pr-submit.yml 
          pip install -r requirements.txt
```
8. Profit!

PS: Please note, creating a PR for this will take less time than writing above steps in detail however since you are new to codebase, even if you take a week or more it's fine, no hurries and worries ;)

**Comment by leelavg** [comment]
@slayer321 is it possible to provide an ack if you've picked up this issue?

**Comment by slayer321** [comment]
Thank you @leelavg  for explaining to me everything in detail.  I am new to k8 and opensource so I got overwhelmed with so many things. Currently, I'm learning docker and how all things work that you have mentioned. So I think I might take time. But once I'm done learning. I will intimidate you and will start working on this issue.
Sorry for the late reply.

**Comment by leelavg** [comment]
@slayer321 ack, thanks for the confirmation. 

No problem, one step at a time. Do comment here for any help required.

@vatsa287 do you see this issue as a good-first, need some confirmation 😅.

**Comment by vatsa287** [comment]
@leelavg Thanks for the steps in detail, yes it is a good first issue now😀

**Comment by leelavg** [comment]
@slayer321 didn't hear from you so went ahead and raised PR to fix this
1. You can look at the merged PR #592 to get to know how above comment is implemented
2. You can look around other issues if you are interested in taking up any
